src/common/hooks/useGetSetting.ts

```diff
@@ -0,0 +1,52 @@
+/**
+ * Invoice Ninja (https://invoiceninja.com).
+ *
+ * @link https://github.com/invoiceninja/invoiceninja source repository
+ *
+ * @copyright Copyright (c) 2022. Invoice Ninja LLC (https://invoiceninja.com)
+ *
+ * @license https://www.elastic.co/licensing/elastic-license
+ */
+
+import { Client } from '../interfaces/client';
+import { Settings } from '../interfaces/company.interface';
+import { useGroupSettingsQuery } from '../queries/group-settings';
+import { useCurrentCompany } from './useCurrentCompany';
+
+export function useGetSetting() {
+  const company = useCurrentCompany();
+
+  const { data: groupSettings } = useGroupSettingsQuery({ perPage: 1000 });
+
+  return (client: Client | undefined, propertyKey: keyof Settings) => {
+    if (groupSettings && client) {
+      if (client.settings[propertyKey] !== undefined) {
+        return client.settings[propertyKey];
+      }
+
+      if (
+        client.group_settings &&
+        client.group_settings.settings[propertyKey] !== undefined
+      ) {
+        return client.group_settings.settings[propertyKey];
+      }
```

Comment:
i think this block should be below the next client settings check? otherwise client.group_settings will override client.settings?

Reply:
@turbo124 Hmm, no. So, the logic for `client.settings` is above this logic on lines `23-24`. Then, this part of the logic is the second one when we are checking the `group_settings` object (Group Settings level) to see if the object exists in the Client object.

The logic below this one is when the `group_settings` object does not exist but the `group_settings_id` prop is populated, so we need to resolve the group_settings object from the pulled list, and finally, we have fallback on the company settings level.

Therefore, I'm sure that this does not override the Client settings level, the `client.settings` logic is just above. Let me know your thoughts.

Reply:
I'm not sure why you have the wrapping groupSettings && client... because

if client.settings.prop !== undefined..... return  it immediately.

Reply:
> I'm not sure why you have the wrapping groupSettings && client... because
> 
> if client.settings.prop !== undefined..... return it immediately.

@turbo124 Yeah, so the reason for wrapping the whole logic first in two simple conditions for client && groupSettings is pretty straightforward.

I wrapped it in the `client` part of the condition so we do not have to do it in the logic where we call the function. We will probably always need to check `if client !== undefined` in the logic where the function is called because we are aware that the result of the function will not be what we EXPECT if the client is `undefined`. We know that we will get the first value from the company or group level settings for sure.

The second condition is related to the groupSettings array, which is fetched from the API or pulled from the cache. So, let's imagine this scenario: we are expecting to get the value from this function, and it needs to be from the group settings level, but the object `group_settings` does NOT exist in the passed client object. So, basically, if we do not have this check, the first value which will be returned from the function will be incorrect because there is a case for sure when the API call for groupSettings is not finished, and the function is returning the company level value.

So basically, we are returning the value only when we are sure that it should be the correct one or let's say when we have the client object !== undefined and the groupSettings array pulled from the API. Otherwise, we have those checks for edge cases that might happen.

The second thing you mentioned is checking the value if it is undefined or not `(if client.settings.prop !== undefined)`. The reason we cannot return it immediately is simply because if I return it like this `return client.settings.prop`, the logic will always return what is on the client settings level, it will never go to the next level to check if there is a populated value.

There is only immediate return for the company level, which is the last one (`return company.settings.prop`), because we are aware that there are no other checks after that and we can return the value. Without if statements in the first three cases, we will always get only the first one returned.